### PR TITLE
Make `ContinualLearningExperiment` compatible with task & class scenarios

### PR DESCRIPTION
**Pull request by karangrewal**
Here, I make `ContinualLearningExperiment` compatible with the "task" and "class" training paradigms in continual learning (see here for details: https://arxiv.org/abs/1904.07734). The major changes are:

(a)`target_transform` is applied directly to the dataset targets after the dataset is instantiated in `setup_experiment` (note that passing target_transform in the config (which is the current setup) gives an error I wasn't able to resolve),

(b) `train_model` and `evaluate_model` in model utils are now equipped to handle multiple output heads depending on the scenario (see PR #424 for details), and thus they are used.

Implementation for dealing with task & class scenarios was previously added to nupic.research [here](https://github.com/numenta/nupic.research/tree/master/nupic/research/frameworks/continual_learning/multihead), but not integrated with Vernon.

### CODE REVIEW

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py [hunk 1]
```diff
@@ -141,6 +156,162 @@ def compute_task_indices(cls, config, dataset):
...
+            return [label for label in range(
+                self.num_classes_per_task * task_num,
+                self.num_classes_per_task * (task_num + 1)
+            )]
+        elif self.cl_experiment_type == "class":
+            return [label for label in range(
+                self.num_classes_per_task * (task_num + 1)
+            )]
+        elif self.cl_experiment_type == "domain":
+            raise NotImplementedError
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/cl_experiment.py:179]: Maybe add a `#TODO` comment here so it's easy to search the file and see what tasks are left for later.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py [hunk 2]
```diff
@@ -141,6 +156,162 @@ def compute_task_indices(cls, config, dataset):
...
+
+    def train_cl_model(
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/experiments/cl_experiment.py:181]: I'm curious why this function and `evaluate_cl_model` is needed. I believe they may not be. Ideally, we should avoid duplicate code whenever possible. Is it possible to call `set_active_tasks` on the samplers during `run_task`? This would look something like
```
    def run_task(self):
    
        # Configure the sampler to load only samples from classes of the active task.
        active_classes = self.get_active_classes(self.current_task)
        self.train_loader.sampler.set_active_tasks(active_classes)
        self.val_loader.sampler.set_active_tasks(active_classes)

        # Run epochs
        for _ in range(self.epochs):
            self.run_epoch()
```

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py:179]: done.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py:181]: It's already using the sampler to set active tasks, but the task and class scenarios treat output nodes separately when evaluating the loss.

E.g., in splitMNIST, the task scenario computes the output for all 10 units, but only the outputs of task 0 (the output heads for classes 0 & 1) are updated during the backward pass.

File: nupic/research/frameworks/pytorch/cl_model_utils.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+in continual learning scenarios.
+"""
+
+import time
+
+import torch
+import torch.nn.functional as F
+
+
+def train_cl_model(
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/cl_model_utils.py:33]: Do we need these extra functions? If it is just a one-line change to switch the heads, can't this be a callback, or done prior to calling train model?

File: nupic/research/frameworks/pytorch/cl_model_utils.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    :param scenario: continual learning scenario, one of "task", "domain", or "class"
+    :param num_classes_per_task: the number of classes per task
+    """
+    if scenario == "task":
+        return [label for label in range(
+            num_classes_per_task * task_num,
+            num_classes_per_task * (task_num + 1)
+        )]
+
+    elif scenario == "domain":
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/cl_model_utils.py:195]: If we are not implementing, better to remove this. Doesn't seem like we will be working with the domain scenario anytime soon.

**Review by lucasosouza — CHANGES_REQUESTED** [review]
I don't think we need the customized train and validate functions. Switching the heads is a one line change - that can be either a callback or done before calling this function. 

My understanding is the class scenario is already implemented - this PR only implements the task scenario. Please correct me if my understanding is incorrect.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: Just saw this, apologies for the repeated comment below. 

If it is a task scenario, the output will have two units and the target will be one hot encoded to reflect that, so they will match. If it is a class scenario, you only want the distribution to be between the classes which actually show in the dataset. So why not backpropagate from the other nodes? They will never show as target, so we will push them to zero, which is what is supposed to happen in that case. 

Can you please point to a theoretical reference so I can understand this a little bit better.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py (hunk 1 above)
Comment (lucasosouza) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: Also repeated comment below. We have no plans to deal with the domain scenario, I don't see a reason to leave this TODO here.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: It seems the definition comes from this paper: https://arxiv.org/pdf/1904.07734v1.pdf. I will give my understanding below, please correct me if I am wrong, but it might be better if we talk over zoom.

In Task-IL, it is a multiheaded scenario. You are informed of which task it is before hand and you can switch heads, to use a head dedicated to that task. In this PR I don't see where the heads are switched.

In Class-IL, it is a single head. You are not informed of the task beforehand. If you mask the output, that means you have the task information (which is not supposed to be available). My understanding is that in Class-IL, given the information of which task we are in is not available, we have to backpropagate from all nodes.

Comment (lucasosouza) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: A bit more info from the appendix of that paper. 

For Task-IL: " normalization performed by the final softmax layer only takes into account these active nodes, and that learning is thus restricted to those nodes". Masking output when calculating gradients is not enough, softmax will still be computed over all nodes in the forward pass.  

For Class-IL: "With the Class-IL scenario only classes from up to the previous task can be predicted.". This makes sense, and it means the previous implementation is incorrect, since the only active nodes are the ones from classes seen so far (before we were calculating for all nodes). It is also aligned with the code in `get_active_classes` in this PR. However, the softmax calculation is also incorrect in this scenario.

Comment (lucasosouza) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: An easier way to implement this is in Task-IL, keeping track of the heads, and appending the proper head to the network when task is switched. No need to change train or eval. 

In Class-IL, start with the number of heads for the present task, and incrementally add new nodes to the single head as more tasks are presented. As before, can add the new nodes when the task is switched, no need to change train or eval.

File: nupic/research/frameworks/pytorch/cl_model_utils.py (hunk 3 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/cl_model_utils.py:33]: You're right, we can add this one line to `train_model` in model_utils.py, but just beware that function is slowly becoming hard to read/understand since there are so many options ([see here](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/model_utils.py#L37-L50)).

File: nupic/research/frameworks/pytorch/cl_model_utils.py (hunk 4 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/cl_model_utils.py:195]: Ok, I will remove it.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: I'm  basing my implementation off the one in the paper that you linked. Here is how they select output units to compute the loss, just as we do: [link](https://github.com/GMvandeVen/continual-learning/blob/master/encoder.py#L185). Also, since that relies on a variable `active_classes`, here is where they compute it, just like us: [link](https://github.com/GMvandeVen/continual-learning/blob/master/train.py#L77-L82)

Comment (lucasosouza) [reply nupic/research/frameworks/vernon/experiments/cl_experiment.py]: There is no issues on how you filter y or compute active classes - the only issue in this implementation on the PR is how softmax is calculated. 

Another option you can consider is taking softmax out of the network, and filtering the logits before calculating the softmax. That can avoid having to "switch" the heads of the network, but it might require rewriting the train and eval functions, so it will reuse less code.

**Review by lucasosouza — APPROVED** [review]
As discussed, the softmax issue is not there, since is being calculated outside the network during the loss function calculation. But the other remains as discussed in the other PR. I agree is a good intermediate solution, will probably require refactoring to a more permanent one at some point. We need to keep the issues with the current implementation in mind so we don't run into bugs down the road.
